parse_test.go

```diff
@@ -116,3 +116,47 @@ func TestParseFile(t *testing.T) {
 	}
 	assert.Equal(t, expected, cmds)
 }
+
+func TestParseReaderHeredocs(t *testing.T) {
+	dockerfile := `RUN <<EOF
+source $HOME/.bashrc && echo $HOME
+echo "Hello" >> /hello
+echo "World!" >> /hello
+EOF
+`
+	cmds, err := ParseReader(bytes.NewBufferString(dockerfile))
+	assert.Nil(t, err)
+	expected := []Command{
+		Command{
+			Cmd:       "RUN",
+			Original:  dockerfile,
+			StartLine: 1,
+			EndLine:   5,
+			Flags:     []string{},
+			Value:     []string{"<<EOF", "source $HOME/.bashrc && echo $HOME\necho \"Hello\" >> /hello\necho \"World!\" >> /hello\nEOF\n"},
+		},
+	}
+	assert.Equal(t, expected, cmds)
+}
+
+func TestParseReaderHeredocsMultiple(t *testing.T) {
+	dockerfile := `COPY <<FILE1 <<FILE2 /dest
+content 1
+FILE1
+content 2
+FILE2
+`
+	cmds, err := ParseReader(bytes.NewBufferString(dockerfile))
+	assert.Nil(t, err)
+	expected := []Command{
+		Command{
+			Cmd:       "COPY",
+			Original:  dockerfile,
+			StartLine: 1,
+			EndLine:   5,
+			Flags:     []string{},
+			Value:     []string{"<<FILE1", "<<FILE2", "/dest", "content 1\nFILE1\n", "content 2\nFILE2\n"},
```

Comment:
this makes even less sense here  how would one piece together the right things here?

almost think these need to be a completely separate structure?

Reply:
Yes, you are correct and those are very good points. It depends on how deep you want to parse the command. I implemented it this way to have a first, high level approach that supported heredocs by parsing the overall RUN command properly, but keeping the whole heredoc as one big list of lines. I felt this would be a good initial step that avoided making too many changes to your library.

If you feel this is too limited and is more confusing than helpful, this can be improved. The buildkit library did add a new internal structure, Heredocs, and their parsed commands now contain a list of Heredocs. If you feel it is better go and parse things at this level, I'd be happy to do it. I can duplicate what they did, adding a list of objects of Heredocs type to the command structure. It would imply adding that Heredoc sturcture to your library, and modifying the glue code that converts from Go to Python to handle this structure. Each heredoc in the list will be properly parsed and will remove those syntax indicators, since that is how they are now stored in the new buildkit structures.

If you think the best approach is to go there, I'll do it. Not sure if you have any opinions what to put on the Value field in that case, since the parse Heredocs would end up in a separate structure, the list of Heredocs. I could leave it as it is, or leave it as null to indicate that the Heredocs list should be used instead.

Reply:
Not really, though they could be labelled as an oversimplification. I'll update the PR with the new buildkit heredoc structure then.

Reply:
@asottile Ok, updated PR. Now the new Heredoc structure that was created and is used in the buildkit library is also used in this library, so the parsed result contains the list of parsed Heredocs, same as the buildkit library would.